# Task
I decided to treat this as a classification problem by creating a new binary variable affair (did the woman have at least one affair?) and trying to predict the classification for each woman. 
 
Dataset 
 
The dataset I chose is the affairs dataset that comes with Statsmodels. It was derived from a survey of women in 1974 by Redbook magazine, in which married women were asked about their participation in extramarital affairs. More information about the study is available in a 1978 paper from the Journal of Political Economy. 
 

 
Code to loading data and modules: 
 
import numpy as np 
 
import pandas as pd 
 
import statsmodels.api as sm 
 
import matplotlib.pyplot as plt 
 
from patsy import dmatrices 
 
from sklearn.linear_model  
import LogisticRegression from 
sklearn.cross_validation  
import train_test_split from sklearn  
import metrics from 
sklearn.cross_validation  
import cross_val_score 
dta = 
sm.datasets.fair.load_pandas().data 
 
 
 
# add "affair" column: 1 represents having affairs, 0 
represents not dta['affair'] = (dta.affairs > 
0).astype(int) 
 
y, X = dmatrices('affair ~ rate_marriage + age + 
yrs_married + children + \ religious + educ + 
C(occupation) + C(occupation_husb)', 
 
dta, return_type="dataframe") 
 
 
X = X.rename(columns = {'C(occupation)[T.2.0]':'occ_2', 
 
'C(occupation)[T.3.0]':'occ_3', 
 
'C(occupation)[T.4.0]':'occ_4', 
 
'C(occupation)[T.5.0]':'occ_5', 
 
'C(occupation)[T.6.0]':'occ_6', 
 
'C(occupation_husb)[T.2.0]':'occ_husb_2', 
 
'C(occupation_husb)[T.3.0]':'occ_husb_3', 
 
'C(occupation_husb)[T.4.0]':'occ_husb_4', 
 
'C(occupation_husb)[T.5.0]':'occ_husb_5', 
 
'C(occupation_husb)[T.6.0]':'occ_husb_6'}) 
 
y = np.ravel(y) 
 
 

In [1]:
import numpy as np 
 
import pandas as pd 
 
import statsmodels.api as sm 
 
import matplotlib.pyplot as plt 
 
from patsy import dmatrices 
 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics 
from sklearn.model_selection import cross_val_score 

In [2]:
dta = sm.datasets.fair.load_pandas().data 


In [3]:
dta

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0               3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1               3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2               4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3               4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4               5.0  27.0          9.0       1.0        1.0  14.0         3.0   
...             ...   ...          ...       ...        ...   ...         ...   
6361            5.0  32.0         13.0       2.0        3.0  17.0         4.0   
6362            4.0  32.0         13.0       1.0        1.0  16.0         5.0   
6363            5.0  22.0          2.5       0.0        2.0  14.0         3.0   
6364            5.0  32.0          6.0       1.0        3.0  14.0         3.0   
6365            4.0  22.0          2.5       0.0        2.0  16.0         2.0   

      occupation_husb   affairs  
0                 5.0  0.111111  
1                 4.0  3.230769  
2                 5.0  1.400000  
3                 5.0  0.727273  
4                 4.0  4.666666  
...               ...       ...  
6361              3.0  0.000000  
6362              5.0  0.000000  
6363              1.0  0.000000  
6364              4.0  0.000000  
6365              4.0  0.000000  

[6366 rows x 9 columns]

Description of Variables 
 
The dataset contains 6366 observations of 9 variables: 
 
rate_marriage: woman's rating of her marriage (1 = very poor, 5 = very good) 
 
age: woman's age 
 
yrs_married: number of years married 
 
children: number of children 
 
religious: woman's rating of how religious she is (1 = not religious, 4 = strongly religious) 
 
educ: level of education (9 = grade school, 12 = high school, 14 = some college, 16 = college graduate, 17 = some graduate school, 20 = advanced degree) 

 
occupation: woman's occupation (1 = student, 2 = farming/semiskilled/unskilled, 3 = "white collar", 4 = teacher/nurse/writer/technician/skilled, 5 = managerial/business, 6 = professional with advanced degree) 
 
occupation_husb: husband's occupation (same coding as above) 
 
affairs: time spent in extra-marital affairs 

In [4]:
dta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6366 entries, 0 to 6365
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   rate_marriage    6366 non-null   float64
 1   age              6366 non-null   float64
 2   yrs_married      6366 non-null   float64
 3   children         6366 non-null   float64
 4   religious        6366 non-null   float64
 5   educ             6366 non-null   float64
 6   occupation       6366 non-null   float64
 7   occupation_husb  6366 non-null   float64
 8   affairs          6366 non-null   float64
dtypes: float64(9)
memory usage: 447.7 KB


In [5]:
dta['affair'] = (dta.affairs > 0).astype(int)

In [6]:
dta

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0               3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1               3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2               4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3               4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4               5.0  27.0          9.0       1.0        1.0  14.0         3.0   
...             ...   ...          ...       ...        ...   ...         ...   
6361            5.0  32.0         13.0       2.0        3.0  17.0         4.0   
6362            4.0  32.0         13.0       1.0        1.0  16.0         5.0   
6363            5.0  22.0          2.5       0.0        2.0  14.0         3.0   
6364            5.0  32.0          6.0       1.0        3.0  14.0         3.0   
6365            4.0  22.0          2.5       0.0        2.0  16.0         2.0   

      occupation_husb   affairs  affair  
0                 5.0  0.111111       1  
1                 4.0  3.230769       1  
2                 5.0  1.400000       1  
3                 5.0  0.727273       1  
4                 4.0  4.666666       1  
...               ...       ...     ...  
6361              3.0  0.000000       0  
6362              5.0  0.000000       0  
6363              1.0  0.000000       0  
6364              4.0  0.000000       0  
6365              4.0  0.000000       0  

[6366 rows x 10 columns]

In [7]:
y, X = dmatrices('affair ~ rate_marriage + age + yrs_married + children + religious + educ + C(occupation) + C(occupation_husb)', 
 dta, return_type="dataframe")

In [8]:
X = X.rename(columns = {'C(occupation)[T.2.0]':'occ_2', 
 
'C(occupation)[T.3.0]':'occ_3', 
 
'C(occupation)[T.4.0]':'occ_4', 
 
'C(occupation)[T.5.0]':'occ_5', 
 
'C(occupation)[T.6.0]':'occ_6', 
 
'C(occupation_husb)[T.2.0]':'occ_husb_2', 
 
'C(occupation_husb)[T.3.0]':'occ_husb_3', 
 
'C(occupation_husb)[T.4.0]':'occ_husb_4', 
 
'C(occupation_husb)[T.5.0]':'occ_husb_5', 
 
'C(occupation_husb)[T.6.0]':'occ_husb_6'}) 

In [9]:
X

Intercept  occ_2  occ_3  occ_4  occ_5  occ_6  occ_husb_2  occ_husb_3  \
0           1.0    1.0    0.0    0.0    0.0    0.0         0.0         0.0   
1           1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
2           1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
3           1.0    0.0    0.0    0.0    1.0    0.0         0.0         0.0   
4           1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
...         ...    ...    ...    ...    ...    ...         ...         ...   
6361        1.0    0.0    0.0    1.0    0.0    0.0         0.0         1.0   
6362        1.0    0.0    0.0    0.0    1.0    0.0         0.0         0.0   
6363        1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
6364        1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   
6365        1.0    1.0    0.0    0.0    0.0    0.0         0.0         0.0   

      occ_husb_4  occ_husb_5  occ_husb_6  rate_marriage   age  yrs_married  \
0            0.0         1.0         0.0            3.0  32.0          9.0   
1            1.0         0.0         0.0            3.0  27.0         13.0   
2            0.0         1.0         0.0            4.0  22.0          2.5   
3            0.0         1.0         0.0            4.0  37.0         16.5   
4            1.0         0.0         0.0            5.0  27.0          9.0   
...          ...         ...         ...            ...   ...          ...   
6361         0.0         0.0         0.0            5.0  32.0         13.0   
6362         0.0         1.0         0.0            4.0  32.0         13.0   
6363         0.0         0.0         0.0            5.0  22.0          2.5   
6364         1.0         0.0         0.0            5.0  32.0          6.0   
6365         1.0         0.0         0.0            4.0  22.0          2.5   

      children  religious  educ  
0          3.0        3.0  17.0  
1          3.0        1.0  14.0  
2          0.0        1.0  16.0  
3          4.0        3.0  16.0  
4          1.0        1.0  14.0  
...        ...        ...   ...  
6361       2.0        3.0  17.0  
6362       1.0        1.0  16.0  
6363       0.0        2.0  14.0  
6364       1.0        3.0  14.0  
6365       0.0        2.0  16.0  

[6366 rows x 17 columns]

In [10]:
X_train, X_test, y_train, y_test=train_test_split(X,y,test_size=0.25, random_state=0)

In [11]:
X_train.shape, X_test.shape, y_train.shape, y_test.shape


((4774, 17), (1592, 17), (4774, 1), (1592, 1))

In [12]:
log_reg=LogisticRegression()
log_reg.fit(X_train, y_train.values.ravel())

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [13]:
prediction=log_reg.predict(X_test)

In [14]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

In [15]:
confusion_matrix=confusion_matrix(prediction, y_test)
print(confusion_matrix)

[[994 314]
 [106 178]]


In [16]:
classification_report=classification_report(y_test, prediction)
print(classification_report)

              precision    recall  f1-score   support

         0.0       0.76      0.90      0.83      1100
         1.0       0.63      0.36      0.46       492

    accuracy                           0.74      1592
   macro avg       0.69      0.63      0.64      1592
weighted avg       0.72      0.74      0.71      1592



In [17]:
accuracy_score=cross_val_score(log_reg, X, y.values.ravel(), scoring='accuracy', cv=5)
print(accuracy_score)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_

[0.70722135 0.72270228 0.71956009 0.72034564 0.75098193]


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [18]:
print(accuracy_score.mean())

0.7241622590180552
